File: src/utils/extremeWeatherRules.ts
import type { ExtremeEvent, ForecastDay, NewAlert } from '../types/rules.ts'

function sortDays(days: ForecastDay[]): ForecastDay[] {
  return [...days].sort((a, b) => a.date.localeCompare(b.date))
}

function makeEvent(
  pointId: number,
  ruleId: string,
  type: string,
  title: string,
  description: string,
  level: ExtremeEvent['level'],
  startAt: string
): ExtremeEvent {
  return { pointId, ruleId, type, title, description, level, startAt }
}

function toAlert(pointName: string, event: ExtremeEvent): NewAlert {
  return {
    pointId: event.pointId,
    fieldId: null,
    level: event.level,
    message: `[极端天气] ${pointName} - ${event.title}：${event.description}`,
    time: Date.now(),
    handled: false,
    source: 'auto',
    ruleId: event.ruleId,
    chain: 'extreme',
    draft: false
  }
}

export function evaluateForecast(
  pointId: number,
  pointName: string,
  days: ForecastDay[]
): { events: ExtremeEvent[]; alertsToCreate: NewAlert[] } {
  const sorted = sortDays(days)
  const events: ExtremeEvent[] = []

  for (const day of sorted) {
    if (day.tempMax >= 40) {
      events.push(
        makeEvent(
          pointId,
          'extreme_heat_40',
          'high_temperature',
          '极端高温',
          `预报最高气温达到 ${day.tempMax}℃`,
          'critical',
          day.date
        )
      )
    }
    if (day.tempMin <= -5) {
      events.push(
        makeEvent(
          pointId,
          'extreme_frost',
          'frost',
          '霜冻风险',
          `预报最低气温 ${day.tempMin}℃`,
          'high',
          day.date
        )
      )
    }
    if (day.windMax >= 17.2) {
      events.push(
        makeEvent(
          pointId,
          'extreme_wind',
          'gale',
          '大风',
          `预报最大风速 ${day.windMax} m/s`,
          'warning',
          day.date
        )
      )
    }
    if (day.precipMm >= 50) {
      events.push(
        makeEvent(
          pointId,
          'extreme_rain',
          'heavy_rain',
          '暴雨',
          `预报日降水 ${day.precipMm} mm`,
          'high',
          day.date
        )
      )
    }
  }

  for (let i = 0; i <= sorted.length - 3; i++) {
    const window = sorted.slice(i, i + 3)
    if (window.every((d) => d.tempMax >= 38)) {
      events.push(
        makeEvent(
          pointId,
          'extreme_heat_3d',
          'high_temperature',
          '连续高温',
          `连续 3 日最高气温 ≥ 38℃（自 ${window[0].date}）`,
          'warning',
          window[0].date
        )
      )
      break
    }
  }

  const alertsToCreate = events.map((event) => toAlert(pointName, event))
  return { events, alertsToCreate }
}

File: ml-bjj/serving/rules/extreme_weather_rules.py
from __future__ import annotations

from datetime import datetime


def evaluate_forecast(point_id: int, point_name: str, days: list[dict]) -> dict:
    sorted_days = sorted(days, key=lambda item: str(item.get("date") or ""))
    events: list[dict] = []

    def make_event(rule_id: str, type_name: str, title: str, description: str, level: str, start_at: str):
        return {
            "pointId": point_id,
            "ruleId": rule_id,
            "type": type_name,
            "title": title,
            "description": description,
            "level": level,
            "startAt": start_at,
        }

    for day in sorted_days:
        temp_max = float(day.get("tempMax") or 0)
        temp_min = float(day.get("tempMin") or 0)
        wind_max = float(day.get("windMax") or 0)
        precip = float(day.get("precipMm") or 0)
        date = str(day.get("date") or "")
        if temp_max >= 40:
            events.append(
                make_event(
                    "extreme_heat_40",
                    "high_temperature",
                    "极端高温",
                    f"预报最高气温达到 {temp_max}℃",
                    "critical",
                    date,
                )
            )
        if temp_min <= -5:
            events.append(
                make_event(
                    "extreme_frost",
                    "frost",
                    "霜冻风险",
                    f"预报最低气温 {temp_min}℃",
                    "high",
                    date,
                )
            )
        if wind_max >= 17.2:
            events.append(
                make_event(
                    "extreme_wind",
                    "gale",
                    "大风",
                    f"预报最大风速 {wind_max} m/s",
                    "warning",
                    date,
                )
            )
        if precip >= 50:
            events.append(
                make_event(
                    "extreme_rain",
                    "heavy_rain",
                    "暴雨",
                    f"预报日降水 {precip} mm",
                    "high",
                    date,
                )
            )

    for index in range(0, max(0, len(sorted_days) - 2)):
        window = sorted_days[index : index + 3]
        if window and all(float(item.get("tempMax") or 0) >= 38 for item in window):
            events.append(
                make_event(
                    "extreme_heat_3d",
                    "high_temperature",
                    "连续高温",
                    f"连续 3 日最高气温 ≥ 38℃（自 {window[0].get('date')}）",
                    "warning",
                    str(window[0].get("date") or ""),
                )
            )
            break

    alerts = [
        {
            "pointId": event["pointId"],
            "fieldId": None,
            "level": event["level"],
            "message": f"[极端天气] {point_name} - {event['title']}：{event['description']}",
            "time": int(datetime.now().timestamp() * 1000),
            "handled": False,
            "source": "auto",
            "ruleId": event["ruleId"],
            "chain": "extreme",
            "draft": False,
        }
        for event in events
    ]
    return {"events": events, "alertsToCreate": alerts}
